test/16-ports.spec.js

```diff
@@ -0,0 +1,378 @@
+import { expect } from './lib/chai-config';
+import { PortMap } from '../src/compose/ports';
+
+describe('Ports', function() {
+	describe('Port string parsing', function() {
+		it('should correctly parse a port string without a range', function() {
+			expect(new PortMap('80')).to.deep.equal(
+				new PortMap({
+					internalStart: 80,
+					internalEnd: 80,
+					externalStart: 80,
+					externalEnd: 80,
+					protocol: 'tcp',
+					host: '',
+				}),
+			);
+
+			expect(new PortMap('80:80')).to.deep.equal(
+				new PortMap({
+					internalStart: 80,
+					internalEnd: 80,
+					externalStart: 80,
+					externalEnd: 80,
+					protocol: 'tcp',
+					host: '',
+				}),
+			);
+		});
+
+		it('should correctly parse a port string without an external range', () =>
+			expect(new PortMap('80-90')).to.deep.equal(
+				new PortMap({
+					internalStart: 80,
+					internalEnd: 90,
+					externalStart: 80,
+					externalEnd: 90,
+					protocol: 'tcp',
+					host: '',
+				}),
+			));
+
+		it('should correctly parse a port string with a range', () =>
+			expect(new PortMap('80-100:100-120')).to.deep.equal(
+				new PortMap({
+					internalStart: 100,
+					internalEnd: 120,
+					externalStart: 80,
+					externalEnd: 100,
+					protocol: 'tcp',
+					host: '',
+				}),
+			));
+
+		it('should correctly parse a protocol', function() {
+			expect(new PortMap('80/udp')).to.deep.equal(
+				new PortMap({
+					internalStart: 80,
+					internalEnd: 80,
+					externalStart: 80,
+					externalEnd: 80,
+					protocol: 'udp',
+					host: '',
+				}),
+			);
+
+			expect(new PortMap('80:80/udp')).to.deep.equal(
+				new PortMap({
+					internalStart: 80,
+					internalEnd: 80,
+					externalStart: 80,
+					externalEnd: 80,
+					protocol: 'udp',
+					host: '',
+				}),
+			);
+
+			expect(new PortMap('80-90:100-110/udp')).to.deep.equal(
+				new PortMap({
+					internalStart: 100,
+					internalEnd: 110,
+					externalStart: 80,
+					externalEnd: 90,
+					protocol: 'udp',
+					host: '',
+				}),
+			);
+		});
+
+		it('should throw when the port string is incorrect', () =>
+			expect(() => new PortMap('80-90:80-85')).to.throw);
+	});
+
+	describe('toDockerOpts', function() {
+		it('should correctly generate docker options', () =>
+			expect(new PortMap('80').toDockerOpts()).to.deep.equal({
+				exposedPorts: {
+					'80/tcp': {},
+				},
+				portBindings: {
+					'80/tcp': [{ HostIp: '', HostPort: '80' }],
```

Comment:
I had to convert all of these to strings to match the types - I'm assuming that's correct but just want to double check